### ISSUE TITLE
fixed_mask is INERT at decode: sample() ignores it entirely (blocks the necklace regen)

### ISSUE THREAD

**Issue by maraxen** [maraxen/PrxteinMPNN#110]
Found 2026-07-15 by the two-armed differential, on its **first run**, during the 260715 control-knob audit (PR #108). This is the deepest layer of the epic's root finding and it **blocks S6/S7 (the 882-bead necklace regen)**.

## The finding

`sample()` accepts `fixed_mask`/`fixed_tokens`, threads them all the way to the decode — and **the output is unaffected**.

Decisive probe (real `proteinmpnn_v_48_020`, `tests/data/1ubq.pdb`):

```python
mask = np.ones(512, dtype=np.float32)                          # freeze EVERY position
tok  = np.full(512, MPNN_ALPHABET.index('W'), dtype=np.int32)  # ...to Tryptophan
r = sample(inputs=['tests/data/1ubq.pdb'], checkpoint_id='proteinmpnn_v_48_020',
           num_samples=1, temperature=0.5, random_seed=0,
           fixed_mask=mask, fixed_tokens=tok)
```

Result — first 60 residues of the output:

```
MTIFVKFEDGETLELEVSPSDTIAELKEQIEKQTGIPPEEQVLIYRGKVLKPDKTLADYN
fraction of first 76 residues == W:  0.0
```

**Zero Trp.** A normal ubiquitin-like design. Every position was declared frozen to W and not one was held.

Two-armed check, same seed: arm A (masked) and arm B (unmasked) are **byte-identical**.

## What IS verified to work (so the bug is narrow)

- `SamplingSpecification(fixed_mask=...)` → `spec.run_spec.sampling.fixed_mask` is set.
- `_prepare_fixed_controls` **is called** (instrumented) and returns the correct mask: `nonzero=3`, `shape=(1,512)` for a triad mask.
- `kernel_dispatch.py:216-227` passes **both** `fixed_mask=fm` and `fixed_tokens=ft` into `build_inference_bundle`.
- `bundle_builder.py:237-238` sets them on the bundle.
- `decode/autoregressive.py:200-205` initialises the sequence from `cond.fixed_mask > 0.5` → `cond.fixed_tokens`, and `:341-343` does the group-fixed substitution.

Every hop looks right on inspection. **The effect is still absent.** So either a different decode path runs than the one read above, or the value is lost between the bundle and the sampling loop. Not yet traced — flagging now rather than sitting on it.

## Why this matters

`--fixed-policies catalytic_triad` was decorative because the planner never wrote the mask into the spec (fixed in PR #108, `dd6618c`). **That fix is correct but NOT SUFFICIENT.** The mask now reaches the model — and the model still ignores it. **The regen would still produce nothing fixed.**

## Why nobody caught it

The obvious check — *"the catalytic triad came back native"* — **passes vacuously**. The triad is the most conserved thing in the protein; an unmasked ProteinMPNN emits H/D/C anyway. `assert_fixed_held` (tev_design `analyze_sampling_grid.py:208`) is exactly that shape and would have been green on the void library.

The **two-armed** differential is what caught it, and it caught it immediately:

> Same bead, same seed. Arm A masked, arm B unmasked. Assert (1) A has zero entropy at masked positions; (2) **B has nonzero entropy at ≥1 masked position**; (3) **A ≢ B**.

Clause (3) fired instantly, with no biology needed. Clause (2) is the load-bearing one nobody had written. **A decorative knob makes A ≡ B byte-for-byte** — and that is a fact a completion gate can never see. 882/882 is a statement about processes exiting, not about arms differing.

## Next steps

1. Trace which decode function `sample()` actually reaches (`make_inference_plan` → `plan.decode_fn`; `host/plan.py:449-455` branches on `STEDecode`). Confirm whether `decode/autoregressive.py:200-205` is live at all for this path.
2. Land the two-armed differential as a permanent gate. It is the only check that would have caught this, and it is cheap.
3. **S6/S7 are blocked until the mask actually holds a residue.** Do not launch the regen; it would reproduce the exact failure it exists to fix.

Reproduce with the snippet above — it takes ~30s on CPU.

**Comment by maraxen** [comment]
## Root cause found, and the scope is narrower than this issue's title says.

**Root cause (confirmed by instrumentation, not just reading):** `host/plan.py::make_inference_plan` (lines 842-848) resolves `decode_fn` with exactly two branches:

```python
if getattr(spec, "sampling_strategy", None) == "straight_through":
    decode_mode = STEMode()
else:
    decode_mode = ConditionalMode()
```

There is no branch that ever constructs `AutoregressiveMode()`. So `decode_fn` is **always** `ConditionalDecode` (`inference/decode/conditional.py`) for the default `sampling_strategy="temperature"`. Traced live with `jax.debug.print`/monkeypatching against the exact repro in this issue:

```
decode_fn_type: ConditionalDecode
sequence_oh_sum: 0.0
```

`ConditionalDecode.__call__` does a single teacher-forced forward pass (`model.decoder.call_conditional`, via `_decode_one_step`) conditioned on `cond.sequence_oh` (all zeros here — `bundle_builder.py:234`) and `cond.ar_mask`, then argmaxes. It reads `cond.sequence_oh`, `cond.ar_mask`, `cond.bias`, `cond.tie_group_map`, `cond.state_position_map` — **never** `cond.fixed_mask`/`cond.fixed_tokens`. `AutoregressiveDecode.__call__` (the class that *does* read them, at :200-205 and :340-346, and which I independently re-verified is itself correct via `tests/inference/decode/test_autoregressive.py::test_fixed_n_to_c_causal_conditioning`) is **never instantiated** for this call path — confirmed by the trace: it never fires.

So this isn't a dropped value at some hop. It's structural: `sample()` (and `inspect()`/`jacobian()`, which share this same `make_inference_plan`) never runs the decode class that has any mechanism for honoring `fixed_mask` at all, regardless of how correctly the value is threaded through every upstream hop.

## Scope correction: the necklace campaign's real bead-sampling path is NOT affected

This issue's repro calls `sample()` directly with a single structure. That is **not** what the necklace campaign actually runs for its k=4-state PoE beads. `host/campaign.py:1087-1094` routes any row with `len(inputs) > 1` to `sample_multistate_poe_campaign_row` → `sample_multistate_poe_bead` → `sample_autoregressive.kernel` (`sampling/multistate_poe.py`) instead of `sample()`. That path builds `AutoregressiveMode()` directly and bypasses `make_inference_plan` entirely.

I ran the same two-armed differential (real `proteinmpnn_v_48_020`, 1ubq.pdb ×2 as a 2-state bead, `max_length=32` to keep the wave-scan compile cheap) against `sample_multistate_poe_bead` directly:

```
[AR call] fixed_mask_nonzero=8.0 fixed_tokens_head=[18 18 18 18 18]
Arm A (masked), 4 samples, first 8 positions: all W (18), every sample, zero entropy
fraction == W in arm A masked region: 1.0
Arm B (unmasked), same region: fraction == W: 0.03125 (background rate)
Arm A byte-identical to Arm B: False
```

**The mask holds correctly on the real PoE path.** No existing test covered this combination (`tests/sampling/test_multistate_poe.py` has zero references to `fixed_mask`/`fixed_tokens`), so this was genuinely unverified before now, not just under-tested.

## Net effect on tev_design's necklace regen

- **PoE beads (90% of designs per bead, per `build_necklace_p2_manifest.py`'s `SPIKE_IN_FRACTION`)**: fixed_mask works. **Not blocked by this issue.**
- **Spike-in beads (10% of designs per bead, single-structure, `len(inputs)==1`)**: these go through the broken `sample()` path. If they're meant to hold the same triad constraint as their parent PoE bead (which "fold-in spike-in control" implies they should, to be a valid control), they are still affected by this issue.
- Separately (not this issue, filing as tev_design debt): `build_necklace_p2_manifest.py:276` still passes `--fixed-policies catalytic_triad` to `aminx campaign plan`, which is now a hard error against this branch's `--fixed-arm` rework (PR #108) — the script needs updating regardless before any regen run.

## Updated next steps

1. **Fix the root cause**: add the missing `AutoregressiveMode()` branch to `make_inference_plan` (or decide `sample()`'s single-shot `ConditionalDecode` path is intentional and rename/reframe it — but then `fixed_mask`/`fixed_tokens` should be rejected loudly for that path rather than silently accepted and ignored).
2. **Decide the spike-in question**: do spike-in rows need the triad held? If yes, they need either (a) the `make_inference_plan` fix, or (b) routing spike-in rows through the PoE path too (`sample_multistate_poe_bead` with `n_states=1`, if that's even valid — needs checking).
3. Land the two-armed differential (both variants used here) as a permanent regression gate for whichever path(s) get fixed.
4. `inspect()`/`jacobian()` are also affected by the same root cause (they share `make_inference_plan`) — separate from the necklace regen, but worth noting since P1 (aminx#issue TBD, this session's alphabet audit) already found unrelated bugs on those same two commands.

Reproduction scripts available on request — both are small (<100 lines), instrumented with monkeypatches on `ConditionalDecode.__call__`/`AutoregressiveDecode.__call__` to make the decode-class dispatch and conditioning values observable at runtime rather than inferred from source reading.

**Comment by maraxen** [comment]
## Escalation: this is bigger than fixed_mask — `ConditionalDecode`/`sample()` is fully deterministic

Follow-up to the root-cause comment above. While auditing whether other knobs (bias, omit_AA, temperature) have the same problem, I tested whether `ConditionalDecode` samples stochastically at all — it doesn't, and this is now confirmed cleanly (initial runs were confounded by `truncation_strategy="random_crop"` firing on a too-small `max_length`; re-verified at `max_length=128`, ≥ the test structure's native length, no cropping):

```
num_samples=4, single call:        all 4 samples byte-identical
random_seed ∈ {0,1,2,3}:            all 4 byte-identical
temperature ∈ {0.01, 0.5, 1.5}:      all 3 byte-identical
identical call, twice:               byte-identical (reproducible, at least)
```

**`sample()` for any single-structure call returns the exact same sequence regardless of `num_samples`, `random_seed`, or `temperature`.** This matches the earlier root cause exactly: `ConditionalDecode.__call__` has no `jax.random.categorical`/any RNG-driven draw anywhere in it or in `decode/_kernel.py::_decode_one_step` — it's a single teacher-forced forward pass, argmaxed. `fixed_mask` was the first symptom found; `temperature`/`random_seed`/`num_samples`-diversity are the same root cause, not separate bugs.

**Practical impact, beyond what's already noted**: any single-structure ("spike-in") campaign row has never produced diverse designs — every "sample" drawn for a given structure is byte-for-byte the same sequence. This is on top of (not instead of) the fixed_mask finding: spike-in beads were never fixing the triad *and* were never actually sampling more than one distinct sequence per structure in the first place.

This doesn't change the fix recommendation (add the missing `AutoregressiveMode` branch to `make_inference_plan`, or make `ConditionalDecode`'s current behavior an explicit, correctly-named non-sampling mode that rejects `num_samples>1`/non-trivial `temperature` rather than silently accepting and ignoring them) — it just means the blast radius on the sampling path (not just fixed_mask) is larger than initially scoped.

**Comment by maraxen** [comment]
## Fixed: `make_inference_plan` never had an `AutoregressiveMode` branch

Root cause confirmed exactly as traced in the comments above. `host/plan.py::make_inference_plan` resolved `decode_fn` with only two branches — `STEMode()` for `sampling_strategy="straight_through"`, else always `ConditionalMode()`. There was no third branch. `sample()` always got `ConditionalDecode` (a single teacher-forced pass), which is why `fixed_mask`, `temperature`, `random_seed`, and `num_samples`-diversity were all inert — not because any one of them was individually mishandled, but because the decode class that would have honored them was never built.

**Fix**: added `resolve_decode_mode(run_spec, *, purpose)` as the single place this decision is made — both `make_inference_plan` and `sampling/multistate_poe.py::sample_multistate_poe_bead` (the PoE path) now call it, instead of each independently deciding. `purpose` (`"sample"` vs `"score"`) is the piece that was missing: `sampling_strategy` alone can't distinguish "I want new sequences" from "I want to evaluate a given one" — both a sampling call at its default `"temperature"` strategy and every scoring call share that identical value. `sample()` now passes `purpose="sample"`; `score()`'s averaged path and `inspect()` pass `purpose="score"` and are unaffected.

**Verified live** (real `proteinmpnn_v_48_020`, same repro structure as above, `max_length=128` to avoid an unrelated `truncation_strategy="random_crop"` confound found while testing this):

```
fixed_mask: masked arm 100% held in the frozen region (was: worked on the PoE path only)
unmasked arm: 0% in the same region (background rate)
num_samples=4: NOT all identical anymore (was: always identical)
random_seed ∈ {0,1,2}: NOT all identical anymore (was: always identical)
```

All three were confirmed-broken behaviors from the original report and the escalation comment; all three are now fixed together, from the same change.

`sample_multistate_poe_bead` also now consults the same resolver and raises `NotImplementedError` if a caller ever sets `sampling_strategy="straight_through"` on a multi-state spec — that combination was never wired for genuine PoE sampling and remains out of scope, but it's now an honest error instead of the field being silently ignored.

Commits: `2afc0c2` (RunSpec completion), `d0c6cdc` (the resolver + fix). Part of EPIC #1541 P4 (inference-runner/host-sinks), which the epic's own tracking had marked "NOT STARTED" — see the follow-up investigation in this same audit pass for why.

Still open: the permanent Tier-C differential test suite (debt #708) formalizing this verification as committed tests rather than one-off scripts, and the coverage guard preventing a future silent field drop (in progress, same PR).

**Comment by maraxen** [comment]
Merged to main: cb57d5b38fc4a65cf0e62f01eafc22e5d6c70012
